flink-tests/src/test/java/org/apache/flink/test/checkpointing/ChangelogPeriodicMaterializationITCase.java

```diff
@@ -0,0 +1,190 @@
+                            @Override
+                            protected void beforeElement() throws Exception {
+                                if (getRuntimeContext().getAttemptNumber() == 0
+                                        && currentIndex == TOTAL_ELEMENTS / 4) {
+                                    waitWhile(
+                                            () ->
+                                                    completedCheckpointNum.get() <= 0
+                                                            || PENDING_MATERIALIZATION.stream()
+                                                                    .noneMatch(Future::isDone));
+                                    PENDING_MATERIALIZATION.clear();
+                                    throwArtificialFailure();
+                                } else if (getRuntimeContext().getAttemptNumber() == 1
+                                        && currentIndex == TOTAL_ELEMENTS / 2) {
+                                    waitWhile(
+                                            () ->
+                                                    completedCheckpointNum.get() <= 1
+                                                            || PENDING_MATERIALIZATION.stream()
+                                                                    .noneMatch(Future::isDone));
+                                    throwArtificialFailure();
+                                }
+                            }
+                        }));
+    }
+
+    @Test
+    public void testFailedMaterialization() throws Exception {
+        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+        env.enableCheckpointing(10)
+                .enableChangelogStateBackend(true)
+                .getCheckpointConfig()
+                .setCheckpointStorage(TEMPORARY_FOLDER.newFolder().toURI());
+        env.setStateBackend(
+                        new DelegatedStateBackendWrapper(
+                                delegatedStateBackend,
+                                snapshotResultFuture -> {
+                                    RunnableFuture<SnapshotResult<KeyedStateHandle>>
+                                            snapshotResultFutureWrapper =
+                                                    new FutureTaskWithException<>(
+                                                            () -> {
+                                                                SnapshotResult<KeyedStateHandle>
+                                                                        snapshotResult =
+                                                                                snapshotResultFuture
+                                                                                        .get();
+                                                                if (PENDING_MATERIALIZATION.size()
+                                                                        == 0) {
```

Comment:
I've just noticed another issue:
This check is performed **after** `snapshotResultFutureWrapper` is added to `PENDING_MATERIALIZATION`; so it will never evaluate to `true`, and materialization will never fail.

Futhermore, subsequent materialization wrappers can be added to the list, before the async pool picks this up (the interval is only 20ms currntly).

Reply:
I've just noticed another issue:
This check is performed **after** `snapshotResultFutureWrapper` is added to `PENDING_MATERIALIZATION`; so it will never evaluate to `true`, and materialization will never fail.

Futhermore, subsequent materialization wrappers can be added to the list, before the async pool picks this up (the interval is only 20ms currntly).